tiledb/sm/rest/rest_profile.h

```diff
@@ -0,0 +1,181 @@
+  /* ****************************** */
+  /*       PARAMETER DEFAULTS       */
+  /* ****************************** */
+
+  /** The default name of a RestProfile. */
+  static const std::string DEFAULT_NAME;
+
+  /** The user's REST password. */
+  static const std::string DEFAULT_PASSWORD;
+
+  /** The namespace that should be charged for the request. */
+  static const std::string DEFAULT_PAYER_NAMESPACE;
+
+  /** The user's REST token. */
+  static const std::string DEFAULT_TOKEN;
+
+  /** The default address for REST server. */
+  static const std::string DEFAULT_SERVER_ADDRESS;
+
+  /** The user's REST username. */
+  static const std::string DEFAULT_USERNAME;
+
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /**
+   * Constructor.
+   *
+   * @param name The name of the RestProfile. Defaulted to "default".
+   */
+  RestProfile(const std::string& name = RestProfile::DEFAULT_NAME);
+
+  /**
+   * Constructor. Intended for testing purposes only, to preserve the user's
+   * $HOME path and their profiles from in-test changes.
+   *
+   * @param name The name of the RestProfile.
+   * @param homedir The user's $HOME directory, or desired in-test path.
+   */
+  RestProfile(const std::string& name, const std::string& homedir);
+
+  /** Destructor. */
+  ~RestProfile() = default;
+
+  /* ****************************** */
+  /*              API               */
+  /* ****************************** */
+
+  inline std::string name() {
+    return name_;
+  }
+
+  /**
+   * Sets the given parameter to the given value.
+   *
+   * @param param The parameter to set.
+   * @param value The value to set on the given parameter.
+   */
+  void set(const std::string& param, const std::string& value);
+
+  /**
+   * Retrieves the value of the given parameter.
+   *
+   * @param param The parameter to fetch.
+   * @return The value of the given parameter.
+   */
+  std::string get(const std::string& param) const;
+
+  /** Saves this profile to the local file. */
+  void save();
+
+  /** Removes this profile from the local file. */
+  void remove();
+
+  /**
+   * Exports this profile's parameters and their values to a json object.
+   *
+   * @return A json object of this RestProfile's parameter : value mapping.
+   */
+  json to_json();
+
+  /**
+   * Dumps the parameter : value mapping in json object format.
+   *
+   * @return This RestProfile's parameter : value mapping in json object format.
+   */
+  std::string dump();
+
+ private:
+  /* ********************************* */
+  /*            PRIVATE API            */
+  /* ********************************* */
+
+  /** Loads the profile parameters from the local json file, if present. */
+  void load_from_json_file(const std::string& filename);
+
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** The version of this class. */
```

Comment:
I am a bit confused by what version_ is. Is this just because we anticipate making changes to the format of this file?

Reply:
Yes, that's right. It was touched upon very briefly in the design doc, though I don't think we ever discussed it at length.